README.md
<div align="center" style="position: relative; width: 800px; height: 400px; overflow: hidden;">
  <img src="https://github.com/awcator/awcatorX/assets/54628909/e10f7800-3ac3-4c14-8435-0990623521ce" alt="Logo" style="position: absolute; top: 50%; left: 50%; transform: translate(-50%, -50%); width: 100%; height: 100%; object-fit: cover;">
</div>

# awcatorX

A little 64bit Kernel writing in C as a hobby project.
This is still a WIP project, and I'm still learning.

# Todo

| Task                                 | Reference                                                                        | Status  |
|--------------------------------------|----------------------------------------------------------------------------------|---------|
| Custom BootLoader                    |                                                                                  | ongoing |
| Grub based booting                   |                                                                                  | To Do   |
| UEFI support with multiboot2 support | [Multiboot Specification](https://en.wikipedia.org/wiki/Multiboot_specification) | To Do   |
| helloWorld msg                       |                                                                                  | To Do   |
| VGA/HDMI display support             |                                                                                  | To Do   |
| KeyboardSupport                      |                                                                                  | To Do   |
| MouseSupport                         |                                                                                  | To Do   |
| GUI support                          |                                                                                  | To Do   |
| DiskSupport                          |                                                                                  | To Do   |
| FileSystemSupport                    |                                                                                  | To Do   |
| USB support                          |                                                                                  | To Do   |
| Rebuilding with custom language      |                                                                                  | To Do   |
| GPU support                          |                                                                                  | To Do   |
| GlibC support                        |                                                                                  | To Do   |
| Virtualization                       |                                                                                  | To Do   |

# How it works

[See Documentation](documentation/0.how_it_works.md)

# My setup

ArchLinux x64, i5 4cores 2HW threaded, IDE: Clion
Debugging tools: qemu-system-x86_64, gdb, [Yurichev's Calculator](https://yurichev.com/progcalc/)

# References to start "Hello world Kernel"

https://wiki.osdev.org/Bare_Bones  <br>
https://www.gnu.org/software/grub/manual/multiboot2/multiboot.html#Example-OS-code <br>
https://github.com/mattmassicotte/kernel-tutorial/tree/master <br>
https://github.com/TralahM/simple-os-kernel/blob/master/kernel.cpp <br>
https://github.com/mig-hub/mikeOS <br>
https://dev.to/frosnerd/writing-my-own-boot-loader-3mld <br>
https://github.com/cfenollosa/os-tutorial <br>
https://github.com/klange/toaruos/releases/tag/0.5.0-rel <br>
https://github.com/vinsdragonis/NoobOS <br>
https://github.com/vinsdragonis/Wyvern <br>
http://www.osdever.net/bkerndev/Docs/basickernel.htm <br>
https://github.com/asido/OS <br> G(OLD)
https://github.com/pdoane/osdev <br> G(OLD)
https://github.com/flaneur2020/fleurix <br>
https://github.com/chenyukang/Panda <br>
https://github.com/MQuy/mos/tree/master/examples <br>

# References to post HelloWorld

https://wiki.osdev.org/Creating_an_Operating_System <br>
https://github.com/piot/tiny-libc <br>
https://github.com/soumitradev/assembly-fun <br>
https://github.com/no92/uefi-bare-bones https://wiki.osdev.org/User:No92/UEFI_Bare_Bones <br>
http://www.sebastianmihai.com/snowdrop/src/ <br>
https://github.com/pdoane/osdev <br>

# Inspiration

https://github.com/AlgorithMan-de/wyoos <br>
https://github.com/klange/toaruos/releases?page=10 <br>
https://github.com/LemonOSProject/LemonOS/tree/0.1.0 <br>


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. <br>

## [Unreleased] AwcatorX 0.0.1

Work is going on. It's a hobby project. I'll commit in random days. Its more of low-level reading than coding for me
yet.

### Added

- SimpleBootLoader, Simple Kernel

### Changed

### Fixed

-----------------------
Theme taken from: https://gist.github.com/juampynr/4c18214a8eb554084e21d6e288a18a2c <br>

-----------------------
Maintainer and Dev: https://github.com/Awcator

src/bootloader/disk-access.asm
; Two ways to Read, LBA(logical block address) and CHS (cylinder head sector)  way

; In LBA if u want to read at 58376 byte
; LBA= 58376/512  ==114where 512 is the sector size of the disks. HDD usually 512, DVD/CD above 2048b or 2KB
; then from block, we need to get the offset to reach the data pointer
; offset= 58376    %   512=8
; 512*114+8=58376

; CHS way
; BIOS provider 13h to read from disk in realMode https://www.ctyme.com/intr/int.htm
; http://www.ctyme.com/intr/rb-0607.htm
DISK_READ:
    mov ah, 0x2 ;DISK - READ SECTOR(S) INTO MEMORY operator
    mov al, 0x1 ;number of sectors to read
    mov ch, 0x0 ;our cylinder number
    mov cl, 0x2 ; sector number to read
    mov dh, 0x0 ; head number
    ;mov dl, someDriveNumber, when I boot from qemu as single disk, bios automatically loads up my drive number at dl. ill just make use of the same. 0x80 was my DL.
    ;  (0 = floppy, 1 = floppy2, 0x80 = hdd, 0x81 = hdd2)
    ;   eg. qemu -fda boot_sect_main.bin. read from floppy-disk
    mov bx,buffer; output is stored at ES:BX -> data buffer. So we bring our memory space address into bx
    int 0x13    ; call the interrupt to read the DISK
    jc error    ; if any error while reading disk it will put carry flag set to 1, if carry=1?jump to error. else print out disk contents
    mov si, buffer
    call print_string_pointed_by_ds_and_si
    jmp CODE    ; we dont want execute next line. so go back to where it was called

; say error if error occurred while reading
error:
    mov si, disk_load_error
    call print_string_pointed_by_ds_and_si ; from bootloader.asm

src/apps/boot_sector_apps/boot_pong.asm
; Incomplete just tried to do it. game size exceeded 512Bytes. so parked for now. At least basic graphics works
ORG 0x7C00          ;   Set the origin to 0x7C00 where the bootloader will be loaded.
BITS 16             ;   Tell assembler we want to use 16bits instructions. We want to run in real mode
;   Some Global Constants
VIDEO_MEMORY equ 0xb800     ;   VGA color memory starting address
ROW_LENGTH_BYTES equ   160  ;   80 characters*2Bytes=160bytes for each line
PLAYER_X_AXIS equ 4         ;   Draw player at 2nd letter of a line. why not 1? sometimes not visible to eye due to the edge of screen. so move to 2nd letter(*2 bytes for color+letter).
CPU_X_AXIS equ 154          ;   160-2char space*2 byte size

;   Setup video mode
mov ax,0x03         ;   a text mode with 80x25 chars and 16 colors VGA, 0x01 is a monochrome
int 10h             ;   setup video-mode. Now screen blackout

;   Point es to videoMemory
mov ax,VIDEO_MEMORY       ;   if it is a monochrome, memory starts from 0xb000
mov es,ax                 ;   es:di =0xb800:0x0000

game_loop:
    ;   clear screen every cycle. or repaint whole screen with black     https://wiki.osdev.org/Text_UI
    xor ax,ax       ;   set color and text to print as empty. Black pixels(letter)
    xor di,di       ;   reset di to 0 and increment all the way till cx
    mov cx, 80*25   ;   total number of times to draw color (80 rows*25 col) pixel screen
    rep stosw       ;   same as mov [es:di],ax; inc di. replaced by 1 byte instruction. Basically it writes into video memory with black

    xor di,di
    mov ax,0xDE41   ;   write letter A (0x41) in 0xDE color
    stosw           ;   put ax contents to es:di position. that is indirectly writing into video memory results in drawing a letter


    ;   Draw middle line/separator
    xor al,al        ;   clear up al (the letter). otherwise it will print letter with along the line
    mov ah, [drawColor]     ;   white bg, black fg
    mov di, 78       ;   80 columns we hv . 80/2=40. index start from 0. so 40-1=39. we will draw line at 39th colum. since each character takes 2 bytes, 39*2=78. so plot at address 78
    mov cx, 13       ;   draw dash every line by line, giving one space line gap. we hv 25 rows. so 13 times
    .draw_middle_LineLoop:
        stosw
        ;   di is portioned at 78, which is middle of first line, we want to jump to next line everytime to draw line on next line. each line has 80letters, each letters has 2b (letter+color info) and we want to jump every two lines. so its
        ;   so di should be incremented by 80box*2Bytes/box*2 (twice)=320. but stosw by default increment by 1 word that is 2byte, we subtract it.
        add di,ROW_LENGTH_BYTES*2-2
        loop .draw_middle_LineLoop ; until cx times

    ;Draw Player
    imul di,[PLAYER_Y_AXIS],ROW_LENGTH_BYTES    ; position the di to draw player to his y-axis line(row)
    add di,PLAYER_X_AXIS                        ; adjust his x-axis
    mov cl,5
    .drawPlayer_loop:
        stosw                   ; loops until cx
        add di,ROW_LENGTH_BYTES-2
        loop .drawPlayer_loop

    ;Draw CPU:
    imul di,[PLAYER_Y_AXIS],ROW_LENGTH_BYTES    ; position the di to draw player to his y-axis line(row)
    mov cl,5
    .drawCPU_loop:
        mov [es:di+CPU_X_AXIS],ax  ; same as player code, just eliminated stosw. just to show it works same
        add di,ROW_LENGTH_BYTES
        loop .drawCPU_loop

    ;Draw Ball:
    imul di,[BALL_Y_AXIS],ROW_LENGTH_BYTES    ; position the di to draw player to his y-axis line(row)
    add di, [BALL_X_AXIS]
    mov word [es:di], 0x2000 ;   no character with green color



    ;   Cause a simple delay
    ;   https://wiki.osdev.org/Memory_Map_(x86)#:~:text=Note%3A%20the%20EBDA%20is%20a,be%20found%20using%20significantly%20more.
    mov bx,[0x046C] ; get the value of IRQ0 timer ticks since boot, increment twice, wait till it reaches same, that way we cause delay
    inc bx
    inc bx
    .delay_loop:
        cmp [0x046C],bx
            jl .delay_loop  ; jump if less than to back to loop

jmp game_loop
drawColor: db 0xf0
PLAYER_Y_AXIS: dw 10    ; start from 10th line
CPU_Y_AXIS: dw 10    ; start from 10th line
BALL_X_AXIS: dw 66   ; middle character position is 78(the center line), just place few space before it. like 6 space? 6*2=12. 78-12=66
BALL_Y_AXIS: dw 7    ;  little above the user
; boot-sector padding
times 510-($ - $$) db 0 ;   Pad zeros till 510th byte of the binary file
dw 0xAA55               ;   Intel LittleEndian format. Write 0x55AA from 511th byte of the binary file. Now bios thinks this binary as bootloader binary. DW is 2 bytes, so it will write 0x55aa in 511th byte and 512th byte


src/bootloader/custom_interrupt_vector-table.asm
    ;      +-----------------+
    ;      |      IVT        |
    ;      | (Interrupt      |
    ;      |  Vector Table)  |
    ;      +-----------------+
    ;      |      ...        |   <- Lower Memory Addresses
    ;      +-----------------+
    ;      |   IVT Entry 0   |   <- 0x0000:0000 (Size: 4 bytes). 2Bytes points the offset, next 2Bytes points the data segment
    ;      +-----------------+
    ;      |   IVT Entry 1   |   <- 0x0000:0004 (Size: 4 bytes)
    ;      +-----------------+
    ;      |      ...        |
    ;      +-----------------+
    ;      |   IVT Entry 255 |   <- 0x0000:03FC (Size: 4 bytes)
    ;      +-----------------+
    ;      |      ...        |
    ;      +-----------------+
    ; We have total 256 IVT entries, each IVT entries of 4B, 256*4=1KB memory is reserved for IVT

; Define Interrupt. Simple Interrupt that prints 'Z'
handle_zero_interrupt:
    mov ah, 0eh
    mov al, 'Z'
    mov bx, 0x00
    int 0x10
    retf

src/bootloader/bootloader.asm
;    +-------------------------+
    ;|    Interrupt Vector     |        // check interrupt.asm
    ;|    Table (IVT)          |
    ;|  0x00000 - 0x000FF      |
    ;|        (1KB)            |
    ;+-------------------------+
    ;|      BIOS Data Area     |
    ;|  0x00100 - 0x001FF      |
    ;|        (1KB)            |
    ;+-------------------------+
    ;|    Empty / Reserved     |
    ;|  0x00200 - 0x07BFF      |
    ;|       (29KB)            |
    ;+-------------------------+
    ;|    Bootloader Code &    |
    ;|     Data (RAM/ROM)      |
    ;|  0x07C00 - 0x07E7F      |
    ;|       (0.5KB)           |
    ;+-------------------------+
    ;|   Free Memory Area for  |
    ;|    Bootloader Usage     |
    ;|  0x07E80 - 0x09FFF      |
    ;|        (8KB)            |
    ;+-------------------------+
    ;|    Empty / Reserved     |
    ;|  0x0A000 - 0x0EFFFF     |
    ;|       (56KB)            |
    ;+-------------------------+
    ;|     Video Memory        |
    ;|   (Memory-mapped I/O)   |
    ;|  0x0F000 - 0x0FFFF      |
    ;|        (4KB)            |
    ;+-------------------------+
    ;|   Empty / Reserved      |
    ;|  0x100000 - ...         |
    ;+-------------------------+

; The code-design should be done like as if this runs in Intel 8086 processor from 1970s, it should be in 16bit instructions, also called real-mode.
; References for this code should be done by reading 8086 processor architecture but not x86_64 architecture  (x86_64/x86 instructions wont work)
; Author Awcator
; todo https://stackoverflow.com/questions/9660315/bootloader-strange-behavior?rq=3 https://wiki.osdev.org/El-Torito
; https://github.com/asido/OS/tree/master/boot
ORG 0x7C00          ;   Set the origin to 0x7C00 where the bootloader will be loaded.
BITS 16              ;   Tell assembler we want to use 16bits instructions. We want to run in real mode
; Satisfy BIOS parameter block: read https://wiki.osdev.org/FAT#BPB_.28BIOS_Parameter_Block.29
jmp short start
nop
times 33 db 0       ; let the BIOS fill the first 33bytes after nop instruction, since bios likes to feed BIOS param block info to the bootloaders memory section. Not all BIOS fills it

start:              ;   section label, denoted by $$
    ;cli             ;   clear interrupts/Disable interrupts, since we want to  modify segment register, we dont want interrupts to happen by bios
    ;   mov ax, 0x7c0
    ;   mov ds,ax   ; Manually assigns ourself values to the data segments register, bcz we made org as 0x0
    ;   mov es,ax   ; Manually assigns ourself values to the Extra segments register
    ;   mov ax,0x0000 ; should we start ss from 0? or from 0x7c0
    ;   mov ax, 0x00 ; 0x7c0+544
    ;   mov ss,ax   ; Manually assigns Stack segments register
    ;   mov sp, 0x7c00  ; stackPointer above
    ;sti             ; enable back bios interrupts

    ;%include "./src/bootloader/custom_interrupt_vector-table.asm"
    ;mov ax, cs
    ;mov ds, ax
    ;mov word [ds:0x00], handle_zero_interrupt          ;Point 0x00 to our custom code
    ;mov word [ds:0x02], cs                             ;point our code-segment from where code should be picked-up
    ;int 0                                              ;Since we registered our code at 0x00 we will trigger it by calling 0th interrupt

    ;%include "./src/bootloader/disk-access.asm"
    ;jmp DISK_READ;

    CODE:mov si, bootloader_title ;   point sourceIndex register point message address. This register will be used by lodsb
    call print_string_pointed_by_ds_and_si ; ds*16+si
    jmp $
    print_string_pointed_by_ds_and_si:
        mov bh,0xa0        ;   color background/foreground.
        mov bl,0x07
        for_each_letter:
            lodsb         ;   moves letter (byte) pointed by DS:SI to al
            cmp al,0x0    ;   check if letter is null character or 0
                je endof_for_each_letter ; if al==0? then ZeroFlag=1, je checks ZF is 1 or not. if 1 jumps to endforeachLetter
            call ask_bios_to_print_at_AL ; else call bios to print what AL contains
            jmp print_string_pointed_by_ds_and_si ; again for next character, lodsb increments char position
        endof_for_each_letter:
            ret
    ask_bios_to_print_at_AL: ; it assumes al register is already filled with a character to be displayed
        mov ah, 0x0e    ;   or make it 0eh both are same. It acts as argument for bios to do, where  0x0e=VIDEO - TELETYPE OUTPUT operation http://www.ctyme.com/intr/rb-0106.htm
        int 0x10        ;   Ask BIOS to do it by calling interrupt. Why exact value 0x10? http://www.ctyme.com/intr/rb-0106.htm
        ret
; modification of binary file
; we know our bootloader should have signature 0xAA55 so we hv to create the binary with that ending signature
; we want to put 0xAA55 in 511th byte in the binary. So we pad with zeros or anything (better zeros) from current location upto 510th byte location
; how we do is, we will calculate number of bytes to be padded. '$; gives address of last instruction. '$$' gives  the address of 'start' section
disk_load_error: db 'Error loading the DISK',0x0 ; A global variable like definition
bootloader_title: db 'Loaded AwcatorXBootLoader 0.1',0x0 ; A global variable like definition
times 510-($ - $$) db 0 ;   Pad zeros till 510th byte of the binary file
dw 0xAA55               ;   Intel LittleEndian format. Write 0x55AA from 511th byte of the binary file. Now bios thinks this binary as bootloader binary. DW is 2 bytes, so it will write 0x55aa in 511th byte and 512th byte
buffer:                 ;   In case you want for extra purpose. like loading disk or kernel